tests/lang/s02_core/s99_coroutines/t02_signature_restrictions_1.nim

```diff
@@ -0,0 +1,13 @@
+discard """
+"""
+
+## Except for `openArray` and `var` types, there are no restrictions on the
```

Comment:
It's the `var` parameter restriction because we don't know how to handle them in the environment? Because technically we can just treat it as a `ptr` to the that location (out parameter), although perhaps we're inheriting this limitation from closure iterators (haven't poked at those much).

Reply:
Disallowing `var` is not so much a limitation with closure iterators, rather it's for safety. Capturing a `var` as a `ptr` is indeed possible, but then it's very easy to create a situation of dangling pointers. Consider:
```nim
proc coro(x: var int) {.coroutine.} =
  x = 1
  echo x

proc test(): Coroutine[void] =
  var y = 1
  result = launch coro(y)

# upon returning from `test`, the `y` local is gone, so reading/writing from/to
# `x` would be a use-after-(stack)-free  
resume(test())
```

Callsite escape analysis doesn't help, since the coroutine could save the instance in some global, for example.

Reply:
Oh, so we couldn't require `var T` params be treated as `lent T`?

Reply:
Due to `ref`s shared-owner ship semantics, and `Coroutine` beings `ref`, it would not work, yeah.

If `Coroutine` were a unique-ownership type (e.g., something akin to C++'s `std::unique_ptr`), then storing the `var T` params as `lent T` *could* work, but the borrow checker would need to understand such unique-ownership type.

Reply:
Hmm, even if a coroutine callable was `proc coro(self: sink Coroutine): Coroutine[T]`, `sink` doesn't work because it _might_ be consumed (not strong enough) and that doesn't play well with the fact that we need to return it. I guess we'd have to have `{.noalias.}` become first class, as opposed to `restrict` in codgen.

Reply:
I don't think it's a good idea, but we *could* introduce a `CoroutinePtr`, which is a heap-allocated object that has value semantics. Making this type as ergonomic to use as a `ref` type would be quite tricky, however.

Reply:
I agree, I don't think it'd be a good idea. If anything that capability should develop separately (guided by a few more use cases) and if successful used here. Otherwise, I'm guessing there are better solutions.